src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/WriteHelper.cs

```diff
@@ -440,6 +440,33 @@ public async Task<int> CopyValidatorActiveSetHistory(ICollection<ValidatorActive
         return entities.Count;
     }
 
+    public async Task<int> CopyValidatorUptime(ICollection<ValidatorUptime> entries, CancellationToken token)
+    {
+        if (!entries.Any())
+        {
+            return 0;
+        }
+
+        await using var writer = await _connection.BeginBinaryImportAsync("COPY validator_uptime (id, validator_entity_id, from_state_version, epoch_number, epoch_start, epoch_end, proposals_made, proposals_missed) FROM STDIN (FORMAT BINARY)", token);
+
+        foreach (var e in entries)
+        {
+            await writer.StartRowAsync(token);
+            await writer.WriteAsync(e.Id, NpgsqlDbType.Bigint, token);
+            await writer.WriteAsync(e.ValidatorEntityId, NpgsqlDbType.Bigint, token);
+            await writer.WriteAsync(e.FromStateVersion, NpgsqlDbType.Bigint, token);
+            await writer.WriteAsync(e.EpochNumber, NpgsqlDbType.Bigint, token);
+            await writer.WriteAsync(e.EpochStart, NpgsqlDbType.TimestampTz, token);
+            await writer.WriteAsync(e.EpochEnd, NpgsqlDbType.TimestampTz, token);
```

Comment:
Open question: maybe we should introduce new table `epoch` with `epoch_number, epoch_start_date, epoch_end_date, validators, ...`?

Reply:
hmm we could but i'm not sure if it is not better to have denormalized read model here. Are we using that in any other place?